cirq-core/cirq/devices/thermal_noise_model.py

```diff
@@ -0,0 +1,251 @@
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Dict, List, Optional, Sequence, Union
+from scipy.linalg import expm
+import numpy as np
+
+from cirq import devices, ops, protocols, qis
+from cirq.devices.noise_utils import (
+    PHYSICAL_GATE_TAG,
+)
+
+if TYPE_CHECKING:
+    import cirq
+
+
+def _left_mul(mat: np.ndarray) -> np.ndarray:
+    """Superoperator associated with left multiplication by a square matrix."""
+    mat = np.asarray(mat)
+    if mat.shape[-1] != mat.shape[-2]:
+        raise ValueError(
+            f'_left_mul only accepts square matrices, but input matrix has shape {mat.shape}.'
+        )
+    dim = mat.shape[-1]
+
+    return np.kron(mat, np.eye(dim))
+
+
+def _right_mul(mat: np.ndarray) -> np.ndarray:
+    """Superoperator associated with right multiplication by a square matrix."""
+    mat = np.asarray(mat)
+    if mat.shape[-1] != mat.shape[-2]:
+        raise ValueError(
+            f'_right_mul only accepts square matrices, but input matrix has shape {mat.shape}.'
+        )
+    dim = mat.shape[-1]
+
+    return np.kron(np.eye(dim), np.swapaxes(mat, -2, -1))
+
+
+def _lindbladian(left_op: np.ndarray) -> np.ndarray:
+    r"""Superoperator representing a Lindbladian.
+
+    The Lindbladian generated by a single operator A is the superoperator
+
+        $$
+        L(\rho) = A \rho A^\dagger - 0.5 (A^\dagger A \rho + \rho A^\dagger A)
+        $$
+
+    Args:
+        left_op: The operator acting on the left in the Lindbladian (A above).
+
+    Returns:
+        Superoperator corresponding to the Lindbladian.
+    """
+    left_op = np.asarray(left_op)
+    right_op = left_op.conj().T
+    square = right_op @ left_op
+    out = _left_mul(left_op) @ _right_mul(right_op)
+    out -= 0.5 * (_left_mul(square) + _right_mul(square))
+    return out
+
+
+def _decoherence_matrix(
+    cool_rate: float, dephase_rate: float, heat_rate: float = 0.0, dim: int = 2
+) -> np.ndarray:
+    """Construct a rate matrix associated with decay and dephasing.
+
+    The units of the matrix match the units of the rates specified.
+    This matrix can be used to construct a noise channel after rescaling
+    by an idling time (to make it dimensionless).
+
+    Args:
+        cool_rate: Decay rate of the system, usually 1 / T_1
+        dephase_rate: Static dephasing rate of the system, usually 1 / T_phi
+        heat_rate: Heating rate of the system (default 0).
+        dim: Number of energy levels to include (default 2).
+
+    Returns:
+        np.ndarray rate matrix for decay and dephasing.
+    """
+    # Heating (related to a^dag)
+    rate_matrix = np.diag(np.arange(1, dim) * heat_rate, 1).T.astype(float)
+    # Cooling (related to a)
+    rate_matrix += np.diag(np.arange(1, dim) * cool_rate, 1)
+    # Dephasing (related to n=a^dag * a)
+    # We specify i^2 since we take square root to get the Lindblad op later.
+    rate_matrix += np.diag(dephase_rate * np.arange(dim) ** 2)
+    return rate_matrix
```

Comment:
If we are doing diagonal only operations it might be a little more efficient to do:

`out[np.diagonal_indices(out.shape[-1])] += np.arange(1, dim) * cool_rate`

or something. Does this impact performance ?

Reply:
Punting on this; optimizations are tracked in #4768.